experimental/Schemes/RationalMap.jl

```diff
@@ -535,70 +544,108 @@ function pullback(phi::RationalMap, C::AbsAlgebraicCycle)
       if !isone(J)
         JJ = IdealSheaf(X, domain(psi), gens(J))
-        comps[JJ] = C[I]
-        success = true
+        return JJ
         break
       end
     end
+  end
+  return nothing
+end
 
-    success && break
-
+function _try_randomized_pullback(phi::RationalMap, I::IdealSheaf)
+  X = domain(phi)
+  Y = codomain(phi)
+  scheme(I) === Y || error("ideal sheaf not defined on the correct scheme")
+  # Find a patch in Y on which this component is visible
+  all_V = [V for V in affine_charts(Y) if !isone(I(V))]
+
+  @show all_V
+  min_var = minimum([ngens(OO(V)) for V in all_V])
+  @show min_var
+  all_V = [V for V in all_V if ngens(OO(V)) == min_var]
+  @show all_V
+  deg_bound = minimum([maximum([total_degree(lifted_numerator(g)) for g in gens(I(V))]) for V in all_V])
+  @show deg_bound
+  all_V = [V for V in all_V if minimum([total_degree(lifted_numerator(g)) for g in gens(I(V))]) == deg_bound]
+  @show all_V
+  V = first(all_V)
+  @show findfirst(x->x===V, affine_charts(Y))
+
+  all_U = copy(affine_charts(X))
+  function complexity(U::AbsSpec)
+    a = realization_preview(phi, U, V)
+    return maximum(vcat([total_degree(numerator(f)) for f in a], [total_degree(denominator(f)) for f in a]))
+  end
+  sort!(all_U, lt=(x,y)->complexity(x)<complexity(y))
 
-    success = false
-    for U in all_U
-      psi = random_realization(phi, U, V)
-      U_sub = domain(psi)
+  for U in all_U
+    psi = random_realization(phi, U, V)
+    U_sub = domain(psi)
 
-      J = pullback(psi)(saturated_ideal(I(V)))
-      if !isone(J)
-        JJ = IdealSheaf(X, domain(psi), gens(J))
-        comps[JJ] = C[I]
-        success = true
-        break
-      end
+    J = pullback(psi)(saturated_ideal(I(V)))
+    if !isone(J)
+      JJ = IdealSheaf(X, domain(psi), gens(J))
+      return JJ
+      break
     end
+  end
+  return nothing
+end
 
-    !success && break
-
-    # If that fails, try the hard way
-    for U in all_U
-      psi_loc = realize_maximally_on_open_subset(phi, U, V)
-      # If we are in different components, skip
-      length(psi_loc) > 0 || continue
-      found = 0
-      J = ideal(OO(domain(first(psi_loc))), elem_type(OO(domain(first(psi_loc))))[])
-      cod_ideal = ideal(OO(U), elem_type(OO(U))[])
-      for (k, psi) in enumerate(psi_loc)
-        if dim(cod_ideal) < dim(I)
-          found = 0
-          break
-        end
-        J = pullback(psi)(I(V))
-        if !isone(J)
-          @assert dim(J) == dim(I)
-          found = k
-          break
-        end
-        cod_ideal = cod_ideal + ideal(OO(U), complement_equation(domain(psi)))
+function _pullback(phi::RationalMap, I::IdealSheaf)
+  X = domain(phi)
+  Y = codomain(phi)
+  scheme(I) === Y || error("ideal sheaf not defined on the correct scheme")
+  # Find a patch in Y on which this component is visible
+  all_V = [V for V in affine_charts(Y) if !isone(I(V))]
+
+  @show all_V
+  min_var = minimum([ngens(OO(V)) for V in all_V])
```

Comment:
Clearly you have bigger fish to fry, but let me just point out that code like the following is always inefficient, as you create a vector to store all the `ngens` values, when you really only need to store a single value at any moment.
```suggestion
  min_var = minimum(ngens(OO(V)) for V in all_V)
```